File: StructuralDesignPatterns/JavaScript/Facade.js
// Subsystem 1
class Light {
    turnOn() {
        console.log("Light is ON");
    }

    turnOff() {
        console.log("Light is OFF");
    }
}

// Subsystem 2
class AirConditioner {
    turnOn() {
        console.log("Air Conditioner is ON");
    }

    turnOff() {
        console.log("Air Conditioner is OFF");
    }
}
// Subsystem 3
class Speaker {
    turnOn() {
        console.log("Speaker is ON");
    }

    turnOff() {
        console.log("Speaker is OFF");
    }
}

// Facade
class HomeAutomationFacade {
    constructor(light, ac, speaker) {
        this.light = light;
        this.ac = ac;
        this.speaker = speaker;
    }

    startMovieNight() {
        console.log("Starting Movie Night...");
        this.light.turnOff();
        this.ac.turnOn();
        this.speaker.turnOn();
    }

    endMovieNight() {
        console.log("Ending Movie Night...");
        this.light.turnOn();
        this.ac.turnOff();
        this.speaker.turnOff();
    }
}

// Client Code
const light = new Light();
const ac = new AirConditioner();
const speaker = new Speaker();
const homeAutomation = new HomeAutomationFacade(light, ac, speaker);

homeAutomation.startMovieNight();
// Output:
// Starting Movie Night...
// Light is OFF
// Air Conditioner is ON
// Speaker is ON

homeAutomation.endMovieNight();
// Output:
// Ending Movie Night...
// Light is ON
// Air Conditioner is OFF
// Speaker is OFF

File: StructuralDesignPatterns/Java/Facade.java
// Subsystem 1
class Light {
    public void turnOn() {
        System.out.println("Light is ON");
    }

    public void turnOff() {
        System.out.println("Light is OFF");
    }
}

// Subsystem 2
class AirConditioner {
    public void turnOn() {
        System.out.println("Air Conditioner is ON");
    }

    public void turnOff() {
        System.out.println("Air Conditioner is OFF");
    }
}

// Subsystem 3
class Speaker {
    public void turnOn() {
        System.out.println("Speaker is ON");
    }

    public void turnOff() {
        System.out.println("Speaker is OFF");
    }
}

// Facade
class HomeAutomationFacade {
    private Light light;
    private AirConditioner ac;
    private Speaker speaker;

    public HomeAutomationFacade(Light light, AirConditioner ac, Speaker speaker) {
        this.light = light;
        this.ac = ac;
        this.speaker = speaker;
    }

    public void startMovieNight() {
        System.out.println("Starting Movie Night...");
        light.turnOff();
        ac.turnOn();
        speaker.turnOn();
    }

    public void endMovieNight() {
        System.out.println("Ending Movie Night...");
        light.turnOn();
        ac.turnOff();
        speaker.turnOff();
    }
}

// Client Code
public class Main {
    public static void main(String[] args) {
        Light light = new Light();
        AirConditioner ac = new AirConditioner();
        Speaker speaker = new Speaker();
        HomeAutomationFacade homeAutomation = new HomeAutomationFacade(light, ac, speaker);

        homeAutomation.startMovieNight();
        // Output:
        // Starting Movie Night...
        // Light is OFF
        // Air Conditioner is ON
        // Speaker is ON

        homeAutomation.endMovieNight();
        // Output:
        // Ending Movie Night...
        // Light is ON
        // Air Conditioner is OFF
        // Speaker is OFF
    }
}
